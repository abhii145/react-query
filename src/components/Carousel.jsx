import { useState, useEffect, useCallback } from "react"
import PropTypes from "prop-types"

const Carousel = ({ images }) => {
  const [currentIndex, setCurrentIndex] = useState(0)

  // Function to go to the next image
  const nextImage = useCallback(() => {
    setCurrentIndex((prevIndex) => (prevIndex + 1) % images.length)
  }, [images.length])

  // Function to go to the previous image
  const prevImage = () => {
    setCurrentIndex(
      (prevIndex) => (prevIndex - 1 + images.length) % images.length
    )
  }

  // Automatically move to the next image every 2 seconds
  useEffect(() => {
    const interval = setInterval(nextImage, 2000) // Move to next image every 2 seconds

    // Clear the interval on component unmount
    return () => clearInterval(interval)
  }, [images.length, nextImage])

  return (
    <div className="relative group">
      <div className="overflow-hidden">
        <div
          className="flex transition-transform duration-500"
          style={{
            transform: `translateX(-${currentIndex * 100}%)`,
          }}
        >
          {images.map((image, index) => (
            <div key={index} className="flex-none w-full">
              <img
                src={image}
                alt={`carousel-image-${index}`}
                className="w-full h-[300px] lg:h-[400px] object-contain rounded-lg shadow-lg"
              />
            </div>
          ))}
        </div>
      </div>

      {/* Left and Right buttons with hover effects */}
      <div className="absolute top-1/2 left-0 right-0 flex justify-between transform -translate-y-1/2 px-4 opacity-0 group-hover:opacity-100 transition-opacity duration-300">
        <button
          onClick={prevImage}
          className="bg-black text-white p-3 rounded-full shadow-lg hover:bg-gray-600 opacity-70 hover:opacity-100 transition-all"
        >
          &#8592;
        </button>
        <button
          onClick={nextImage}
          className="bg-black text-white p-3 rounded-full shadow-lg hover:bg-gray-600 opacity-70 hover:opacity-100 transition-all"
        >
          &#8594;
        </button>
      </div>
    </div>
  )
}
Carousel.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
}

export default Carousel
