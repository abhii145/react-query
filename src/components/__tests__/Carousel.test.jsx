import { render, screen, fireEvent } from "@testing-library/react"
import Carousel from "../Carousel"
import { it, expect, describe, vi } from "vitest"

describe("Carousel", () => {
  const images = ["image1.jpg", "image2.jpg", "image3.jpg"]

  it("matches snapshot", () => {
    const { asFragment } = render(<Carousel images={images} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders correctly", () => {
    render(<Carousel images={images} />)
    images.forEach((image, index) => {
      expect(screen.getByAltText(`carousel-image-${index}`)).toBeInTheDocument()
    })
  })

  it("nextImage function works correctly", () => {
    render(<Carousel images={images} />)
    const nextButton = screen.getByText("→")
    fireEvent.click(nextButton)
    expect(screen.getByAltText("carousel-image-1")).toBeInTheDocument()
  })

  it("prevImage function works correctly", () => {
    render(<Carousel images={images} />)
    const prevButton = screen.getByText("←")
    fireEvent.click(prevButton)
    expect(screen.getByAltText("carousel-image-2")).toBeInTheDocument()
  })

  it("automatic image change works correctly", () => {
    vi.useFakeTimers()
    render(<Carousel images={images} />)
    vi.advanceTimersByTime(2000)
    expect(screen.getByAltText("carousel-image-1")).toBeInTheDocument()
    vi.advanceTimersByTime(2000)
    expect(screen.getByAltText("carousel-image-2")).toBeInTheDocument()
    vi.useRealTimers()
  })
})
