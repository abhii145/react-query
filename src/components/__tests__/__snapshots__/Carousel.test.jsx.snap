// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Carousel > matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="relative group"
    role="region"
  >
    <div
      class="overflow-hidden"
    >
      <div
        class="flex transition-transform duration-500"
        style="transform: translateX(-0%);"
      >
        <div
          class="flex-none w-full"
        >
          <img
            alt="carousel-image-0"
            class="w-full h-[300px] lg:h-[400px] object-contain rounded-lg shadow-lg"
            loading="eager"
            src="image1.jpg"
          />
        </div>
        <div
          class="flex-none w-full"
        >
          <img
            alt="carousel-image-1"
            class="w-full h-[300px] lg:h-[400px] object-contain rounded-lg shadow-lg"
            loading="lazy"
            src="image2.jpg"
          />
        </div>
        <div
          class="flex-none w-full"
        >
          <img
            alt="carousel-image-2"
            class="w-full h-[300px] lg:h-[400px] object-contain rounded-lg shadow-lg"
            loading="lazy"
            src="image3.jpg"
          />
        </div>
      </div>
    </div>
    <div
      class="absolute top-1/2 left-0 right-0 flex justify-between transform -translate-y-1/2 px-4 opacity-0 group-hover:opacity-100 transition-opacity duration-300"
    >
      <button
        aria-label="Previous Image"
        class="bg-black text-white p-3 rounded-full shadow-lg hover:bg-gray-600 opacity-70 hover:opacity-100 transition-all"
      >
        ←
      </button>
      <button
        aria-label="Next Image"
        class="bg-black text-white p-3 rounded-full shadow-lg hover:bg-gray-600 opacity-70 hover:opacity-100 transition-all"
      >
        →
      </button>
    </div>
  </div>
</DocumentFragment>
`;
